torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -239,21 +239,33 @@ const WorkerInfo& RpcAgent::getWorkerInfo() const {
 std::shared_ptr<RpcAgent> RpcAgent::currentRpcAgent_ = nullptr;
 
 bool RpcAgent::isCurrentRpcAgentSet() {
-  return currentRpcAgent_ != nullptr;
+  return std::atomic_load(&currentRpcAgent_) != nullptr;
 }
 
 std::shared_ptr<RpcAgent> RpcAgent::getCurrentRpcAgent() {
-  TORCH_INTERNAL_ASSERT(currentRpcAgent_, "Current RPC agent is not set!");
-  return currentRpcAgent_;
+  std::shared_ptr<RpcAgent> agent = std::atomic_load(&currentRpcAgent_);
+  TORCH_INTERNAL_ASSERT(agent, "Current RPC agent is not set!");
+  return agent;
 }
 
 void RpcAgent::setCurrentRpcAgent(std::shared_ptr<RpcAgent> rpcAgent) {
   if (rpcAgent) {
-    TORCH_INTERNAL_ASSERT(!currentRpcAgent_, "Current RPC agent is set!");
+    std::shared_ptr<RpcAgent> previousAgent;
+    // Use compare_exchange so that we don't actually perform the exchange if
+    // that would trigger the assert just below.
+    std::atomic_compare_exchange_strong(
```

Comment:
This is fancy fix! Thanks! Shall we add a link to this page, or cite its text?

http://www.cplusplus.com/reference/atomic/atomic_compare_exchange_strong/